### ISSUE TITLE
Detekt 1.23.8 can't handle Kotlin 2.3.0 metadata

### ISSUE THREAD

**Issue by eygraber** [detekt/detekt#8865]
## Expected Behavior
Detekt 1.23.8 works with Kotlin 2.3.0

## Observed Behavior
I see a ton of false positives across many different rules

## Steps to Reproduce
Update Kotlin to 2.3.0-RC

## Context

When running with debug, I see 100s of these errors:

> error: class 'kotlin.Unit' was compiled with an incompatible version of Kotlin. The actual metadata version is 2.3.0, but the compiler version 2.0.0 can read versions up to 2.1.0.

Using Detekt 2 alpha works.

## Your Environment

* Version of detekt used: 1.23.8
* Version of Gradle used (if applicable): 9.2.1

**Comment by 3flex** [comment]
> Using Detekt 2 alpha works

For that reason I'm closing this - there are unfortunately no updates planned for the 1.x line.

**Comment by eygraber** [comment]
Is there any timeline for v2 planned? Based on the milestone it looks pretty far from being ready, and considering how big Detekt is in the Kotlin ecosystem, it doesn't seem responsible to force users to choose between keeping Kotlin up to date and using an early alpha.

**Comment by AngeloFilaseta** [comment]
We are in a similar situation—Detekt has been blocking us from updating Kotlin for a while now... I understand that migrating in order to support all the new Kotlin features can be challenging though... I hope we can reach a stable 2.0.0 asap

**Comment by cortinico** [comment]
Thanks for your comments

> Is there any timeline for v2 planned? Based on the milestone it looks pretty far from being ready, and considering how big Detekt is in the Kotlin ecosystem, it doesn't seem responsible to force users to choose between keeping Kotlin up to date and using an early alpha.

I'm unsure what are you suggesting here @eygraber 

Releasing further 1.x releases is heavily unpractical as mentioned by @3flex. We started to publish alpha 2.0.0 versions to unblock users. Is there a strong reasons why you can't use it, rather than avoiding a alpha version?

**Comment by eygraber** [comment]
> rather than avoiding a alpha version

That's a big part of it. There are ecosystems around Detekt (convention plugins, rule providers, etc...) that don't want to use an alpha version, which means end users can't either. 

Then there are migration issues. Updating to v2 could be a large undertaking, from issues like https://github.com/detekt/detekt/issues/8611 to having to update the config (which has a ton of breaking changes). There are false positives introduced by v2 that need to be identified and suppressed, and there are many false negatives that were fixed which need to be handled. 

Of course that would have to be addressed at some point anyways, but it just doesn't seem fair to users to block them from updating Kotlin.

I know v2 has been a massive undertaking, and it's been going on for a while, and I'm super appreciative of all the work that goes into Detekt.

**Comment by Maxr1998** [comment]
Unrelated to the rest of the discussion, the fact that Kotlin 2.3.0 is unsupported by detekt 1.X should be made more obvious, for example by pinning this issue.

After updating a few dependencies earlier (including Kotlin), detekt crashed with a NullPointerException. Since I only saw #8901 at first, I had to investigate further and disable the `IgnoredReturnValue` causing the crash, which revealed the false positives and eventually led me to this issue.

**Comment by 3flex** [comment]
That makes sense. Pinned.

**Comment by mymx2** [comment]
```
val gradleKotlinVersion = embeddedKotlinVersion
val detektKotlinVersion = getSupportedKotlinVersion()
val enableDetekt = detektKotlinVersion >= gradleKotlinVersion

tasks.withType<Detekt>().configureEach {
  enabled = enableDetekt
}
```

**Comment by megla-tlanghorst** [comment]
One issue that has come up for us is that the detekt idea plugin is still on 1.X, but we're forced to use the 2.X alpha because of this issue (as JDK 25 support has been introduced with 2.3.0, which is the now current stable LTS version), resulting in issues in different stuff being shown and different settings being used between the IDE and the reports from Gradle.

**Comment by cortinico** [comment]
> One issue that has come up for us is that the detekt idea plugin is still on 1.X, but we're forced to use the 2.X alpha because of this issue (as JDK 25 support has been introduced with 2.3.0, which is the now current stable LTS version), resulting in issues in different stuff being shown and different settings being used between the IDE and the reports from Gradle.

Can you report this here https://github.com/detekt/detekt-intellij-plugin ?

**Comment by hjed** [comment]
Another thing to flag - Spring Boot 3 is end of life in June and Spring Boot 4 requires at least Kotlin 2.2 which is above detekt 1.x.x's supported versions.

**Comment by atulgpt** [comment]
detekt 2.x.x should support kotlin 2.x.x, can you try that and see if that is working?

**Comment by hjed** [comment]
It does but on large codebases it seems to hang fairly often on the current alpha release

**Comment by BraisGabin** [comment]
Yes, that's a know issue. It was fixed on SNAPSHOT. The next alpha version will fix that.

**Comment by hjed** [comment]
Ok cool. Will the next alpha be out soon? My team is already very nervous about using alpha versions, don't think we can switch to snapshot.

**Comment by wbervoets** [comment]
I disabled the IgnoredReturnValue to fix a crash of Detekt using Kotlin 2.3, however it is only due to this ticket that I found out that the cause is Kotlin 2.3. Would it be possible to still fix on 1.23.x branch?

Some other questions:
* Are there other issues running Detekt 1.23.8 on Kotlin 2.3?
* I can not upgrade to non-release versions (so no alfa, betas, RCs)
* Will there be an easy upgrade path from 1.23.8 detekt.yml to 2.x detekt.yml?
* Will there be an easy upgrade path from Gradle 1.23.8 Gradle plugin to Gradle Detekt 2.x plugin?

Thanks!

**Comment by wbervoets** [comment]
Any updates on: 

- Are there other issues running Detekt 1.23.8 on Kotlin 2.3?
- I can not upgrade to non-release versions (so no alfa, betas, RCs)
- Will there be an easy upgrade path from 1.23.8 detekt.yml to 2.x detekt.yml?
- Will there be an easy upgrade path from Gradle 1.23.8 Gradle plugin to Gradle Detekt 2.x plugin?
- Will AGP9.x supported be backported to 1.23.8 ?

Do you have a timeline for a Detekt 2.0 final release?

thanks
Wim
